larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.cc

```diff
@@ -0,0 +1,560 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/TwoViewCosmicRayRemovalTool.cc
+ *
+ *  @brief  Implementation of the two view cosmic removal tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.h"
+
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArMuonLeadingHelper.h"
+#include "larpandoracontent/LArObjects/LArTwoDSlidingFitResult.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewCosmicRayRemovalTool::TwoViewCosmicRayRemovalTool() :
+    m_minSeparation(2.f),
+    m_slidingFitWindow(10000),
+    m_distanceToLine(0.5f),
+    m_minContaminationLength(3.f),
+    m_maxDistanceToHit(1.f),
+    m_minRemnantClusterSize(3),
+    m_maxDistanceToTrack(2.f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::Run(TwoViewDeltaRayMatchingAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
```

Comment:
This function appears to be identical for for each of the removal tools, aside from the type of algorithm that gets passed in. So, I think for the classes that inherit from the RemovalTool, you may want to change the algorithm type to the corresponding NView class they inherit from, then this function can be implemented just once in the RemovalBaseTool, rather than in each of the inheriting classes. For this class, I wonder if this indicates that the inheritance hierarchy isn't quite right? Perhaps the DeltaRayMatrixTool is not needed/might inherit from the RemovalBaseTool/might merge with the RemovalBaseTool? I'm not immediately sure which, if any of these options would be best, but I think there is a lot of overlap between these classes, so that may want further consideration.